### ISSUE TITLE
Optional DataLoader Arguments

### ISSUE THREAD

**Issue by ramppdev** [autrainer/autrainer#16]
In some scenarios it is useful to configure additional dataloader arguments such as the number of workers or pinning memory.
Therefore allow for the specification of `train_loader_kargs` (same for dev and test).

The question is if this should be specified in the main configuration file or in each dataset configuration?
I would suggest the main configuration file, as e.g. the number of workers is also tied to the hardware.

**Comment by ATriantafyllopoulos** [comment]
@AlexanderGbd you recently looked into this, if you want to make a contribution, this is a `tag:good first issue`

**Comment by AlexanderGbd** [comment]
But actually I did not do any changes to the dataloader. I needed a custom sampler for the training loader for which I simply created a custom dataset, where I overwrote the `train_loader` function.
I just initialized the sampler in the `__init__` which I then handed over to this function.

```
class BalancedSamplingDataset(BaseMLClassificationDataset):
    def __init__(
        self,
        ...
    ) -> None:
        super().__init__(
            ...
        )

        ...
        self.sampler = WeightedRandomSampler(weights=self.sample_weights, num_samples=self.num_samples, replacement=True, generator=self._generator)


    @cached_property
    def train_loader(self) -> DataLoader:
        """Get the training loader.

        Returns:
            Training loader.
        """
        return DataLoader(
            self.train_dataset,
            batch_size=self.batch_size,
            shuffle=False, # Mutually exclusive with sampler option
            generator=self._generator,
            collate_fn=self.train_transform.get_collate_fn(self),
            sampler=self.sampler
        )


```

**Comment by ramppdev** [comment]
We currently use properties for our dataloaders:
https://github.com/autrainer/autrainer/blob/68b19a83061f99988dc4a20d6c4aece9911acd58/autrainer/datasets/abstract_dataset.py#L265-L308

This means without changing the API (of the loaders) we would need to pass optional e.g. `train_loader_kwargs` dicts to all dataset `__init__` methods or change the API for our loaders to methods such that we can set the optional arguments there. Then the loader differ from the datasets which are properties though and we loose the advantage of the dataset holding the loader instances, meaning we dont have to manage and use them separately from the dataset instance (see `training.py`).

I think using properties is really intuitive here, and i would love to keep them. However this means that the `__init__` will get increasingly more bloated with tons of optional arguments.
Since we use datasets not only during training (also for e.g. fetching and preprocessing), i have multiple approaches on my mind:
- just bloat the init on its own and not really care about usability
- use the builder pattern to logically structure the different configuration possibilities (possibly with sub builders for different parts)
- switch all properties to methods that manage the appropriate arguments themselves at the expense of passing the management task of the datasets and loaders to the caller side

I want to think long term a bit away from only a YAML based workflow, as our goal is to provide a general API in the future as well and also datasets are central and will be written by users with a high probability. 
The latter two would allow us to get more lightweight setups where we need them. What are your thoughts on this @ATriantafyllopoulos?

**Comment by wagnerp4** [comment]
I propose implementing a `DataLoaderManager` class to handle DataLoader configurations as an additional solution to: 1) bloat init, 2) use builder pattern, 3) methods to replace properties. This solution:

1. Keeps the intuitive property access pattern
2. Allows flexible DataLoader configuration without bloating dataset `__init__`
3. Supports both YAML and programmatic workflows
4. Maintains clean separation of concerns
5. The current training.py usage pattern would remain unchanged since the DataLoaderManager properties return DataLoader instances

# Sample draft implementation:
Changes in file `autrainer/datasets/abstract_dataset.py`:

```python
class DataLoaderManager:
    """Manages DataLoader configurations and instantiation."""
    def __init__(
        self,
        dataset: 'torch.utils.data.Dataset',
        batch_size: int,
        inference_batch_size: Optional[int] = None,
        generator: Optional[torch.Generator] = None,
        collate_fn: Optional[Callable] = None,
    ):
        self.dataset = dataset
        self.batch_size = batch_size
        self.inference_batch_size = inference_batch_size or batch_size
        self.generator = generator
        self.collate_fn = collate_fn
        
        # Default configurations
        self._train_config = {"shuffle": True}
        self._eval_config = {"shuffle": False}
        
    def configure_training(self, **kwargs) -> "DataLoaderManager":
        """Configure training DataLoader parameters."""
        self._train_config.update(kwargs)
        return self
        
    def configure_evaluation(self, **kwargs) -> "DataLoaderManager":
        """Configure evaluation (dev/test) DataLoader parameters."""
        self._eval_config.update(kwargs)
        return self
        
    @property
    def train_loader(self) -> DataLoader:
        """Get training DataLoader."""
        return DataLoader(
            self.dataset,
            batch_size=self.batch_size,
            generator=self.generator,
            collate_fn=self.collate_fn,
            **self._train_config
        )
        
    @property
    def dev_loader(self) -> DataLoader:
        """Get development DataLoader."""
        return DataLoader(
            self.dataset,
            batch_size=self.inference_batch_size,
            generator=self.generator,
            collate_fn=self.collate_fn,
            **self._eval_config
        )
        
    @property 
    def test_loader(self) -> DataLoader:
        """Get test DataLoader."""
        return DataLoader(
            self.dataset,
            batch_size=self.inference_batch_size, 
            generator=self.generator,
            collate_fn=self.collate_fn,
            **self._eval_config
        )
```
Changes in class `AbstractDataset`:

```python
class AbstractDataset(ABC):
    def __init__(
        self,
        path: str,
#
    ) -> None:
# (replacing train_loader, ...)
    @cached_property
    def _train_loader_manager(self) -> DataLoaderManager:
        """Lazily initialize training loader manager."""
        return DataLoaderManager(
            dataset=self.train_dataset,
            batch_size=self.batch_size,
            inference_batch_size=self.inference_batch_size,
            generator=self._generator,
            collate_fn=self.train_transform.get_collate_fn(self)
        )
        
    @cached_property
    def _dev_loader_manager(self) -> DataLoaderManager:
        """Lazily initialize dev loader manager."""
        return DataLoaderManager(
            dataset=self.dev_dataset,
            batch_size=self.batch_size,
            inference_batch_size=self.inference_batch_size,
            generator=self._generator,
            collate_fn=self.dev_transform.get_collate_fn(self)
        )
        
    @cached_property
    def _test_loader_manager(self) -> DataLoaderManager:
        """Lazily initialize test loader manager."""
        return DataLoaderManager(
            dataset=self.test_dataset,
            batch_size=self.batch_size,
            inference_batch_size=self.inference_batch_size,
            generator=self._generator,
            collate_fn=self.test_transform.get_collate_fn(self)
        )
        """Get the test loader manager.

        Returns:
            Test loader manager.
        """
        return self._test_loader_manager

    @property
    def train_loader(self) -> DataLoader:
        """Get the training loader.
        
        Returns:
            Training loader.
        """
        return self._train_loader_manager.train_loader

    @property
    def dev_loader(self) -> DataLoader:
        """Get the development loader.
        
        Returns:
            Development loader.
        """
        return self._dev_loader_manager.dev_loader

    @property
    def test_loader(self) -> DataLoader:
```
<html>
<body>
<html><head></head><body>

Overview:

| Approach | Key Benefits | Key Drawbacks |
|----------|-------------|---------------|
| Bloated Init | Simple implementation, no API changes | Poor maintainability, mixed concerns |
| Builder Pattern | Clean configuration separation, extensible | Complex implementation, YAML complexity |
| Methods | Flexible configuration | Loses caching, breaks existing patterns |
| DataLoaderManager | Maintains properties + flexibility, clean separation | Additional abstraction layer |

Would this work, what do you guys think?

**Comment by ramppdev** [comment]
Hmm yeah, this would definitely work but i think would overcomplicate things even more :D. The best solution is probably to just change the property to a function and collect the overrides as kwargs in the trainer setup. This could even be used to debloat the dataset init a bit and move the batch size there, but would be optional. Let me draft sth up ...
